consensus/src/quorum_store/utils.rs

```diff
@@ -188,159 +215,162 @@ impl ProofQueue {
+        assert!(
+            self.latest_block_timestamp <= block_timestamp,
+            "Decreasing block timestamp"
+        );
+        self.latest_block_timestamp = block_timestamp;
+
+        let peers: Vec<_> = self.author_to_batches.keys().cloned().collect();
+        let mut num_expired_but_not_committed = 0;
+
+        for peer in peers {
+            let mut queue = self.author_to_batches.remove(&peer).unwrap();
+            let num_expired = queue
+                .iter()
+                .take_while(|batch| batch.expiration() < block_timestamp)
+                .count();
+
+            for batch in queue.drain(0..num_expired) {
+                if self
+                    .batch_to_proof
+                    .get(&batch)
+                    .expect("Entry for unexpired batch must exist")
+                    .is_some()
+                {
+                    // non-committed proof that is expired
+                    num_expired_but_not_committed += 1;
+                    if batch.expiration() < block_timestamp {
+                        counters::GAP_BETWEEN_BATCH_EXPIRATION_AND_CURRENT_TIME_WHEN_PULL_PROOFS
+                            .observe((block_timestamp - batch.expiration()) as f64);
+                    }
                 }
+                claims::assert_some!(self.batch_to_proof.remove(&batch));
             }
-        }
-        counters::NUM_TOTAL_TXNS_LEFT_ON_COMMIT.observe(remaining_txns as f64);
-        counters::NUM_TOTAL_PROOFS_LEFT_ON_COMMIT.observe(remaining_proofs as f64);
 
-        (remaining_txns, remaining_proofs)
+            if !queue.is_empty() {
+                self.author_to_batches.insert(peer, queue);
+            }
+        }
+        counters::NUM_PROOFS_EXPIRED_WHEN_COMMIT.inc_by(num_expired_but_not_committed);
     }
 
-    // returns the number of unexpired local proofs
-    pub(crate) fn clean_local_proofs(&mut self, current_block_timestamp: u64) -> Option<u64> {
-        let num_expired = self
-            .local_digest_queue
-            .iter()
-            .take_while(|(_, expiration_time)| *expiration_time < current_block_timestamp)
-            .count();
-        self.local_digest_queue.drain(0..num_expired);
-
-        let mut remaining_local_proof_size = 0;
+    pub(crate) fn num_total_txns_and_proofs(&mut self, current_block_timestamp: u64) -> (u64, u64) {
+        let mut remaining_txns = 0;
+        let mut remaining_proofs = 0;
+        let mut remaining_local_txns = 0;
+        let mut remaining_local_proofs = 0;
 
-        for (digest, expiration) in self.local_digest_queue.iter() {
-            // Not expired. It is possible that the proof entry in digest_proof was already removed
-            // when draining the digest_queue but local_digest_queue is not drained yet.
-            if *expiration >= current_block_timestamp {
-                if let Some(entry) = self.digest_proof.get(digest) {
+        // TODO: if the digest_queue is large, this may be too inefficient
```

Comment:
Can we maintain this on the fly? i.e., update every time we add new proof or commit/expire existing one?